### DISCUSSION TITLE
Scope of RelyingParty object

### DISCUSSION THREAD

**Discussion by muhammadwasay** [Yubico/java-webauthn-server#263] [Q&A]
Is it ok to create one instance of RelyingParty class  at the time application bootstraping and reuse this instance. i.e a Singleton instance for whole java application.
1 - will this instance be threadsafe?
2- is it ok to use the same instance for handling both registeration operations and authentication operations?

What if i create a new RelyingParty instance for every start / finish registeration/authentication operation and then leave it for garbage collection.

**Comment by emlun** [comment]
Hi!

 1. Yes, as noted in the README:

    >No mutable state or side effects - everything (except builders) is thread safe

    So as long as you only reuse the `RelyingParty` instance, not the `RelyingPartyBuilder`, you should be fine.

 3. Yes, that is the intended use.

> What if i create a new RelyingParty instance for every start / finish registeration/authentication operation and then leave it for garbage collection.

That would probably be unnecessarily slow, but should be fine. There's a small amount of [validation logic](https://github.com/Yubico/java-webauthn-server/blob/547389b6c015de01a2247d55a72157689c999d03/webauthn-server-core/src/main/java/com/yubico/webauthn/RelyingParty.java#L415) that runs during `RelyingParty` construction, and the builder itself of course takes a few instructions to finish. I haven't benchmarked it, but I would guess it should take somewhere on the order of 1-10 ms or less to instantiate a `RelyingParty`.
